### PR TITLE
[WOR-61] Pull in azure context for MC workspaces if present

### PR DESCRIPTION
**Pull request by aherbst-broad**
## Context
We will need to surface Azure information for Azure workspaces as we start to introduce these into the UI.
[Relevant ticket](https://broadworkbench.atlassian.net/browse/WOR-61)

## This PR
* Adds a conditional that will reach out to WSM in the event that a workspace is "mc" (aka multi cloud).  
* Pulls in the azure context and adds to the response payload, when present.
   * If azure information is present, this will also suppress the check for GCP bucket details. The premise is that a workspace may not cross clouds.
* Adds unit tests for the new codepath, as well as a unit test for the existing base codepath
------

- [ ] **Submitter**: Include the JIRA issue number in the PR description
- [ ] **Submitter**: Check that the **Product Owner** has signed off on any user-facing changes
- [ ] **Submitter**: Make sure Swagger is updated if API changes
  - [ ] **...and Orchestration's Swagger too!**
- [ ] **Submitter**: If updating admin endpoints, also update [firecloud-admin-cli](https://github.com/broadinstitute/firecloud-admin-cli)
- [ ] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [ ] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * To Demo flag is set
  * Release Summary is filled out, if applicable
  * Add notes on how to QA
- [ ] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [ ] **Submitter**: Database checks:
  * If PR includes new or changed db queries, include the explain plans in the description
  * Make sure liquibase is updated if appropriate
  * If doing a migration, take a backup of the
  [dev](https://console.cloud.google.com/sql/instances/terraform-qfarbdq3lrexxck5htofjs5z6m/backups?project=broad-dsde-dev&organizationId=548622027621)
  and
  [alpha](https://console.cloud.google.com/sql/instances/terraform-r4caezzc35c4tb7pgdhwkmme4y/backups?project=broad-dsde-alpha&organizationId=548622027621)
  DBs in Google Cloud Console
- [ ] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [ ] Tell your tech lead (TL) that the PR exists if they want to look at it
- [ ] Anoint a lead reviewer (LR). **Assign PR to LR**
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [ ] **TL** sign off
- [ ] **LR** sign off
- [ ] **Assign to submitter** to finalize
- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Squash commits and merge to develop
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [ ] **Submitter**: Inform other teams of any API changes via Slack and/or email
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

**Comment by aherbst-broad** [comment]
jenkins retest

File: core/src/test/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceServiceSpec.scala [hunk 1]
```diff
@@ -95,7 +98,7 @@ class WorkspaceServiceSpec extends AnyFlatSpec with ScalatestRouteTest with Matc
     val gcsDAO = Mockito.spy(new MockGoogleServicesDAO("test", googleAccessContextManagerDAO))
     val samDAO = Mockito.spy(new MockSamDAO(dataSource))
     val gpsDAO = new MockGooglePubSubDAO
-    val workspaceManagerDAO = mock[MockWorkspaceManagerDAO](RETURNS_SMART_NULLS)
+    val workspaceManagerDAO = Mockito.spy(new MockWorkspaceManagerDAO())
```
Comment (aherbst-broad) [inline core/src/test/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceServiceSpec.scala:101]: Using a mock here results in a mock with mocks on all methods methods on this class, not an instance of the `MockWorkspaceManagerDAO` supplying the already implemented mock methods. The existing usages of this mock look like they were after verifying calls to the external service, which lends itself to a spy vs a mock.

File: core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala [hunk 2]
```diff
@@ -304,6 +316,42 @@ class WorkspaceService(protected val userInfo: UserInfo,
...
 
+  private def getAzureCloudContextFromWorkspaceManager(workspaceContext: Workspace, parentSpan: Span = null) = {
+    workspaceContext.workspaceType match {
+      case WorkspaceType.McWorkspace => {
+        val span = startSpanWithParent("getWorkspaceFromWorkspaceManager", parentSpan)
+
+        try {
+          val wsmInfo = workspaceManagerDAO.getWorkspace(workspaceContext.workspaceIdAsUUID, userInfo.accessToken)
+
+          Option(wsmInfo.getAzureContext) match {
```
Comment (cahrens) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:327]: What would it mean if the Azure context is not present? I'm wondering if we should do some error handling/logging in that case.

File: core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala [hunk 3]
```diff
@@ -304,6 +316,42 @@ class WorkspaceService(protected val userInfo: UserInfo,
...
+            case Some(azureContext) => Some(WorkspaceAzureCloudContext(
+              azureContext.getTenantId,
+              azureContext.getSubscriptionId,
+              azureContext.getResourceGroupId)
+            )
+            case None => None
+          }
+          // todo pull in GCP context here
```
Comment (cahrens) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala]: I'm not clear what this TODO refers to. Can you provide some more details?

File: core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala [hunk 4]
```diff
@@ -304,6 +316,42 @@ class WorkspaceService(protected val userInfo: UserInfo,
...
+          case e: ApiException =>
+            logger.warn(s"Error retrieving MC workspace from workspace manager [id=${workspaceContext.workspaceId}, code=${e.getCode}]")
+
+            if (e.getCode == StatusCodes.NotFound.intValue) {
+              span.setStatus(Status.NOT_FOUND)
+              throw new RawlsExceptionWithErrorReport(errorReport = ErrorReport(StatusCodes.NotFound, e))
+            } else {
+              span.setStatus(Status.INTERNAL)
+            }
+            throw new RawlsExceptionWithErrorReport(errorReport = ErrorReport(StatusCodes.InternalServerError, e))
```
Comment (cahrens) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala]: Nit: I would put lines 344 and 346 together.

**Review by ungwudik — APPROVED** [review]
lgtm

File: core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala (hunk 2 above)
Comment (aherbst-broad) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:327]: It'd likely mean it's a gcp workspace. At this point in the code's evolution agreed this is an error case, I'll add some handling.

File: core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala (hunk 3 above)
Comment (aherbst-broad) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala]: A WSM workspace can have either a GCP cloud context or an Azure cloud context. For now, all this code path knows about is Azure cloud contexts, but in the long term (once we have a migration plan in place), a GCP context will be retrieved the same way.

That being said, this is explicitly a method to get azure data from WSM, so this comment is confusing, will remove.

File: core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala [hunk 5]
```diff
@@ -304,6 +316,45 @@ class WorkspaceService(protected val userInfo: UserInfo,
...
+          }
+        } catch {
+          case e: ApiException =>
+            logger.warn(s"Error retrieving MC workspace from workspace manager [id=${workspaceContext.workspaceId}, code=${e.getCode}]")
+
+            if (e.getCode == StatusCodes.NotFound.intValue) {
+              span.setStatus(Status.NOT_FOUND)
+              throw new RawlsExceptionWithErrorReport(errorReport = ErrorReport(StatusCodes.NotFound, e))
+            } else {
+              span.setStatus(Status.INTERNAL)
```
Comment (cahrens) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:347]: Why do you want to convert all other codes to INTERNAL? Is there no way to preserve the original status code?

File: core/src/test/scala/org/broadinstitute/dsde/rawls/config/MultiCloudWorkspaceServiceConfigSpec.scala [hunk 6]
```diff
@@ -10,7 +10,9 @@ class MultiCloudWorkspaceServiceConfigSpec extends AnyFlatSpec with Matchers {
   val testConf: Config = ConfigFactory.load()
 
   it should "default to not enabled if no config is present" in {
-    val config = MultiCloudWorkspaceConfig.apply(testConf)
+    val parsed = ConfigFactory.empty()
```
Comment (cahrens) [inline core/src/test/scala/org/broadinstitute/dsde/rawls/config/MultiCloudWorkspaceServiceConfigSpec.scala]: Nit: why not combine lines 13 and 15? I don't find the variable name `parsed` helps me understand the intent.

````
val config = MultiCloudWorkspaceConfig.apply(ConfigFactory.empty())

**Review by cahrens — APPROVED** [review]
I have one question and nit, but I don't necessarily need to re-review.

File: core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala (hunk 5 above)
Comment (aherbst-broad) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:347]: No, these are opencensus specific [codes](https://opencensus.io/tracing/span/status/) that don't directly map to HTTP status codes.

Comment (cahrens) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:347]: Sorry, my comment was more about the next line (348), where we lose the original status code in the `RawlsExceptionWithErrorReport`.

Comment (aherbst-broad) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:347]: Ah, good call, will fix.

File: model/src/main/scala/org/broadinstitute/dsde/rawls/model/WorkspaceModel.scala [hunk 7]
```diff
@@ -651,14 +652,22 @@ case class WorkspaceListResponse(accessLevel: WorkspaceAccessLevel,
...
 case class WorkspaceResponse(accessLevel: Option[WorkspaceAccessLevel],
                              canShare: Option[Boolean],
                              canCompute: Option[Boolean],
                              catalog: Option[Boolean],
                              workspace: WorkspaceDetails,
                              workspaceSubmissionStats: Option[WorkspaceSubmissionStats],
                              bucketOptions: Option[WorkspaceBucketOptions],
-                             owners: Option[Set[String]])
+                             owners: Option[Set[String]],
+                             azureContext: Option[WorkspaceAzureCloudContext]
```
Comment (marctalbott) [inline model/src/main/scala/org/broadinstitute/dsde/rawls/model/WorkspaceModel.scala:669]: We'll need to [publish the new `model` version](https://github.com/broadinstitute/rawls#publish-rawls-model) and update the Rawls and Orchestration swaggers as well

**Review by marctalbott — APPROVED** [review]
Should any of this be in `MultiCloudWorkspaceService` or are we reserving that service exclusively for the `workspaces/mc/...` path?

File: model/src/main/scala/org/broadinstitute/dsde/rawls/model/WorkspaceModel.scala (hunk 7 above)
Comment (aherbst-broad) [reply model/src/main/scala/org/broadinstitute/dsde/rawls/model/WorkspaceModel.scala:669]: @marctalbott is that link correct? It appears to be linking to an unrelated discussion

Comment (marctalbott) [reply model/src/main/scala/org/broadinstitute/dsde/rawls/model/WorkspaceModel.scala:669]: No it isn't... that's what I get for linking without double checking. https://github.com/broadinstitute/rawls#publish-rawls-model is what I meant to link

**Pull request merged**